src/execution/operator/persistent/physical_copy_to_file.cpp

```diff
@@ -72,8 +152,72 @@ SinkResultType PhysicalCopyToFile::Sink(ExecutionContext &context, DataChunk &ch
 		lock_guard<mutex> glock(g.lock);
 		g.rows_copied += chunk.size();
 	}
-	function.copy_to_sink(context, *bind_data, per_thread_output ? *l.global_state : *g.global_state, *l.local_state,
-	                      chunk);
+
+	if (per_thread_output) {
+		auto &gstate = l.global_state;
+		function.copy_to_sink(context, *bind_data, *gstate, *l.local_state, chunk);
+
+		if (file_size_bytes.IsValid() && function.file_size_bytes(*gstate) > file_size_bytes.GetIndex()) {
+			function.copy_to_finalize(context.client, *bind_data, *gstate);
+			gstate = CreateFileState(context.client, *sink_state);
+		}
+		return SinkResultType::NEED_MORE_INPUT;
+	}
+
+	if (!file_size_bytes.IsValid()) {
+		function.copy_to_sink(context, *bind_data, *g.global_state, *l.local_state, chunk);
+		return SinkResultType::NEED_MORE_INPUT;
+	}
+
+	// FILE_SIZE_BYTES is set, but threads write to the same file, need to synchronize finalizing somehow
+	auto &gstate = g.global_state;
+
+	bool take_ownership = false;
+	unique_ptr<GlobalFunctionData> owned_gstate;
+	while (true) {
+		if (owned_gstate) {
```

Comment:
Maybe we can clean this up using a shared/exclusive lock (see `StorageLock` in `src/include/duckdb/storage/storage_lock.hpp`) - the shared lock can be used when calling `copy_to_sink`, the exclusive lock when calling `copy_to_finalize`.